blockingconcurrentqueue.h

```diff
@@ -897,29 +1283,69 @@ class BlockingConcurrentQueue
 	// and at most max.
 	// Using a negative timeout indicates an indefinite timeout,
 	// and is thus functionally equivalent to calling wait_dequeue_bulk.
+  // Also returns 0 if the queue is finished (closed and drained of all remaining elements).
 	// Never allocates. Thread-safe.
 	template<typename It>
 	inline size_t wait_dequeue_bulk_timed(consumer_token_t& token, It itemFirst, size_t max, std::int64_t timeout_usecs)
 	{
 		size_t count = 0;
-		max = (size_t)sema->waitMany((LightweightSemaphore::ssize_t)(ssize_t)max, timeout_usecs);
+		max = (size_t)sema_dequeue->waitMany((LightweightSemaphore::ssize_t)(ssize_t)max, timeout_usecs);
 		while (count != max) {
-			count += inner.template try_dequeue_bulk<It&>(token, itemFirst, max - count);
+			size_t got = inner.template try_dequeue_bulk<It&>(token, itemFirst, max - count);
+      if (got == 0 && closed()) {
+        sema_dequeue->signal();
+        return count;
+      }
+      count += got;
 		}
+		sema_enqueue->signal((LightweightSemaphore::ssize_t)(ssize_t)count);
 		return count;
 	}
 	
 	// Attempts to dequeue several elements from the queue using an explicit consumer token.
 	// Returns the number of items actually dequeued, which can
 	// be 0 if the timeout expires while waiting for elements,
 	// and at most max.
+  // Also returns 0 if the queue is finished (closed and drained of all remaining elements).
 	// Never allocates. Thread-safe.
 	template<typename It, typename Rep, typename Period>
 	inline size_t wait_dequeue_bulk_timed(consumer_token_t& token, It itemFirst, size_t max, std::chrono::duration<Rep, Period> const& timeout)
     {
         return wait_dequeue_bulk_timed<It&>(token, itemFirst, max, std::chrono::duration_cast<std::chrono::microseconds>(timeout).count());
     }
-	
+
+  // Close the queue.
+  // This signals that no more data will be enqueued to dequeue operations.
+  // Further dequeue operations will behave normally until the queue is
+  // completely drained (finished() == true). After that, any dequeue calls
+  // will immediately fail and return.
+  // Callers should check finished() to abort any dequeue loops.
+  // The caller must make sure, that when calling close() all enqueue operations
+  // have finished and no more calls to any enqueue method will be made.
+	// Thread-safe.
+  inline void close()
+  {
+	  bool old_closed = closed_.exchange(true, std::memory_order_relaxed, std::memory_order_relaxed);
+    if (!old_closed) {
+		  sema_dequeue->signal();
+    }
+  }
+
+  // Whether close() has been called on this queue.
+  // When the queue is closed, not enqueue calls must be made.
```

Comment:
Typo. Also, the queue should enforce this (even if it's only in !NDEBUG for performance).

Reply:
I though for quite a while about how to implement this, but TBH I could not come up with a working scheme for this, either using an extra atomic counter for the number of current enqueue operations or just letting the enqueue ops fail if the queue was closed, but I am not sure how to do that reliably either.
In the end I always found some cases where finish() might be true for a short time and that's not acceptable.

Reply:
If wait_enqueue can fail though that could also lead to some weird cases where bulk waiting enqueues insert some but not all elements…